Source/JavaScriptCore/assembler/MacroAssemblerARM64Caps.cpp

```diff
@@ -0,0 +1,545 @@
+/*
+ * Copyright (C) 2013-2019 Apple Inc. All rights reserved.
+ * Copyright (C) 2020,2022 Arm Ltd. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. ``AS IS'' AND ANY
+ * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL APPLE INC. OR
+ * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+ * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+ * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+ * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+ * OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+
+#if ENABLE(ASSEMBLER) && CPU(ARM64_CAPS)  // Morello purecap only.
+#include "MacroAssembler.h"
+
+#include "ProbeContext.h"
+#include <wtf/InlineASM.h>
+
+namespace JSC {
+
+#if ENABLE(MASM_PROBE)
+
+extern "C" void ctiMasmProbeTrampoline();
+
+using namespace ARM64Registers;
+
+#if COMPILER(GCC_COMPATIBLE)
+
+// The following are offsets for Probe::State fields accessed
+// by the ctiMasmProbeTrampoline stub.
+#if CPU(ADDRESS64)
+// The size of a capability.
+#define PTR_SIZE 16
```

Comment:
I did a double take at `PTR_SIZE` being 16 or 4 (and not 8). Given the file is `ARM64Caps` does the `else` clause even need to exist?

Reply:
The `4` case is intended for compressed references (e.g. `ENABLE(JSHEAP_CHERI_OFFSET_REFS)`).

Currently, for purecap, this is only supported up to `tier1asm`, but your comment reminded me to file #12.

Reply:
I wonder if we should add a doc to the `4` case and/or explicitly test that compressed references are enabled?

Reply:
Are we expecting that this will be tested by bors?

Reply:
To that end: https://github.com/capablevms/webkit/issues/13